lib/form/paragraph3.flow

```diff
@@ -630,160 +674,155 @@ keepFormTogether(form : Form) -> bool {
-		lastLine = i == lasti;
+	// TODO: If the line ends with NewLine (which we do not track here), we should consider it a last line and not justify.
+	getLineAlignment = \i -> if (i == lasti)
 		// On the last line, we do not justify.
-		// Also if the line is single with TightWidth, any alignment is equal to StartAlign
-		lineAlignment =
-			if (alignment == Justify() && lastLine) {
-				StartAlign()
-			} else {
-				alignment;
-			}
+		// Also if the line is single with TightWidth, any alignment is equal to StartAlign.
+		GeneralAlignment(alignment with justify=0.0)
+	else
+		alignment;
+
+	optzds = mapi(lines, \i, line -> getOptimizedLine(line.words, getLineAlignment(i)));
+	mtcs = mapi(optzds, \i, optzd -> measureLine(optzd, lines[i].indent));
+	longestLineWidth = fold(mtcs, 0.0, \acc, mtc -> max(acc, mtc.width));
+	width = if (tightWidth) longestLineWidth else max(availableWidth, longestLineWidth);
+	a : ParaLineAcc = foldi(optzds, ParaLineAcc(EmptyList(), 0.0, 0.0), \i : int, acc : ParaLineAcc, optzd : [OptimizedLineElement] -> {
 		f = RenderLine(
-			acc.y, line.words, availableWidth, tightWidth,
-			lineAlignment, line.indent,
+			acc.y, optzd, width, mtcs[i],
+			getLineAlignment(i), lines[i].indent,
 			i, styleTree, rtl, isMarked
 		);
-		nwidth = max(acc.width, f.width);
-		// For interline spacing, it turns out that the old paragraph had crazy behaviour, so we have to double it
-		// to prevent especially label games with explicitly placed things from breaking - REMOVED, NOT NECESSARY
-		nheight = acc.y + f.height + if (lastLine) 0.0 else interlineSpacing;
-		nbaseline = f.baseline;
+		nheight = acc.y + mtcs[i].height + if (i==lasti) 0.0 else interlineSpacing;
 		ParaLineAcc(
-			Cons(Size2(const(zeroWH), f.form), acc.lines),
-			nwidth,
+			Cons(Size2(const(zeroWH), f), acc.lines),
 			nheight,
-			if (i == 0 || !topLineBaseline) nbaseline else acc.baseline
+			if (i == 0 || !topLineBaseline) acc.y+mtcs[i].asc else acc.baseline
 		);
 	});
 
-	Pair(
-		Baseline(const(a.baseline), Size2(const(WidthHeight(if (tightWidth) a.width else max(availableWidth, a.width), a.y)),
-			Group(list2array(a.lines))
-		)),
-		a.width
+	Triple(
+		Baseline(
+			const(a.baseline),
+			Size2(
+				const(WidthHeight(width, a.y)),
+				Group(list2array(a.lines))
+			),
+		),
+		longestLineWidth, a.y
 	);
 }
 
-ParaLineResult(form : Form, width : double, height : double, baseline : double);
+ParaLineMetrics(width : double, height : double, asc : double);
 OptimizedLineElement(f : GhostForm, metrics : Behaviour<FormMetrics>, interactivityIdM : Maybe<int>);
 
-getOptimizedLine(words : [ParaWord], alignment : ParaLineAlignment) -> [OptimizedLineElement]{
-	if (!isBiDiEnabled() && alignment != Justify()) {
+getOptimizedLine(words : [ParaWord], alignment : GeneralAlignment) -> [OptimizedLineElement]{
+	if (!isBiDiEnabled() && equalDoubles(alignment.justify, 0.0)) {
```

Comment:
really bad code here. why we have to compare double instead of just check if it justify?

Reply:
Sure, bool will bring more clearance.